storage/remote/codec.go

```diff
@@ -118,14 +118,19 @@ func FromQuery(req *prompb.Query) (int64, int64, []*labels.Matcher, error) {
 }
 
 // ToQueryResult builds a QueryResult proto.
-func ToQueryResult(ss storage.SeriesSet) (*prompb.QueryResult, error) {
+func ToQueryResult(ss storage.SeriesSet, sampleLimit int) (*prompb.QueryResult, error) {
+	numSamples := 0
 	resp := &prompb.QueryResult{}
 	for ss.Next() {
 		series := ss.At()
 		iter := series.Iterator()
 		samples := []*prompb.Sample{}
 
 		for iter.Next() {
+			numSamples += 1
+			if numSamples > sampleLimit {
+				return nil, fmt.Errorf("too many samples")
```

Comment:
Can we also wrap the error with what is the limit?

Reply:
We should probably have some metrics around this too.

Reply:
And make it return a 400, not a 500...